Allows filtering payments by anonymous donations or supporters

clarissalimab opened the pull request:
This is how it looks:

![anonymous-filter](https://user-images.githubusercontent.com/15739610/135303281-967d16be-49c3-40e3-ab89-3f5593a993e3.gif)

Closes [tix #3261](https://github.com/CommitChange/tix/issues/3261).

Depends on #111 
Depends on #112 
Depends on #113

wwahammy commented:
Just from a usability standpoint, I'm not sure whether that's the best UI. I wonder if it just needs a header or something?

wwahammy commented on lib/query/query_payments.rb:
```diff
@@ -232,6 +232,20 @@ def self.select_to_filter_search(npo_id, query)
 
+    if query[:anonymous].present?
+      expr = if(query[:anonymous] == 'true')
+        expr.where(
+          '(coalesce(supporters.anonymous, false) = $anonymous OR coalesce(donations.anonymous, false) = $anonymous)',
+          anonymous: query[:anonymous]
+        )
```
We'll need to add an index on anonymous for supporters and donations for this; otherwise it'll be way too slow on large nonprofits.

wwahammy commented on app/views/components/tables/filtering/_filter_anonymous.html.erb:
```diff
@@ -0,0 +1,22 @@
+<%- # License: AGPL-3.0-or-later WITH Web-Template-Output-Additional-Permission-3.0-or-later -%>
+<!-- partial: components/filter_anonymous -->
+<form>
+	<fieldset>
+		<input type='radio' name='anonymous' id='radio-anonymous' value='true'>
+    <!--= on 'change' apply_radio_filter -->
+		<label for='radio-anonymous'>Anonymous</label>
+	</fieldset>
+		
+	<fieldset>
+		<input type='radio' name='anonymous' id='radio-not-anonymous' value='false'>
+    <!--= on 'change' apply_radio_filter -->
+    <label for='radio-not-anonymous'>Not-anonymous</label>
+	</fieldset>
+
+	<fieldset>
+		<input class='radio--both' type='radio' name='anonymous' id='radio-either' checked='checked' value=''>
+			<!--= on 'change' apply_radio_filter -->
+		<label for='radio-either'>Any</label>
+	</fieldset> 
+</form>
+<!-- partial end: components/filter_anonymous -->
```
We need to decide on the exact UI but good work as always :)

clarissalimab replied:
Do you think that having:

**Anonymous?**
* Yes
* No
* Either

would be better?

clarissalimab replied:
I'm going to do it and send both versions for wendy to see what they think!

clarissalimab commented:
After incorporating the changes and replacing the nulls for false on my local database, it seems the result came considerably worse :laughing: I didn't save the results from my local database, but we can see it in production after we deploy the other part. This is how it looks *after* the changes:

Not anonymous:
```
 Planning Time: 10.948 ms
 JIT:
   Functions: 140
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 13.647 ms, Inlining 0.000 ms, Optimization 4.892 ms, Emission 81.203 ms, Total 99.741 ms
 Execution Time: 76112.578 ms
```

Anonymous:

```
 Planning Time: 9.086 ms
 Execution Time: 5989.728 ms
```

wwahammy commented:
Hmm, a few ideas:

* is the anonymous field being indexed?
* Are you using the index? If you're doing `COERCE(anonymous, false)`, it won't use the index.

clarissalimab commented:
I did try it with and without index, but I actually thought that we wouldn't be indexing anymore since it's a boolean field. The output was similar, though. I think it's not being used, there's just one part on the output that mentions "index" that I believe is unrelated, but I'm not sure if I'm reading it right. This is the full output for the not anonymous query:

```
 Nested Loop  (cost=358980.23..485695.51 rows=945843 width=93) (actual time=36328.534..71613.290 rows=929505 loo
ps=1)
   ->  Gather Merge  (cost=358979.96..469138.96 rows=945843 width=106) (actual time=36328.460..44076.594 rows=92
9505 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Sort  (cost=357979.93..358965.18 rows=394101 width=106) (actual time=36309.476..37843.120 rows=3098
35 loops=3)
               Sort Key: payments.date DESC
               Sort Method: external merge  Disk: 36104kB
               Worker 0:  Sort Method: external merge  Disk: 36144kB
               Worker 1:  Sort Method: external merge  Disk: 36384kB
               ->  Hash Join  (cost=224685.14..298451.28 rows=394101 width=106) (actual time=30065.106..34689.45
2 rows=309835 loops=3)
                     Hash Cond: (payments.nonprofit_id = nonprofits.id)
                     ->  Parallel Hash Join  (cost=224500.28..297229.15 rows=394101 width=93) (actual time=30018
.865..31598.826 rows=309835 loops=3)
                           Hash Cond: (payments.supporter_id = supporters.id)
                           ->  Parallel Hash Join  (cost=154725.36..213663.72 rows=394101 width=61) (actual time
=22487.659..24080.008 rows=309835 loops=3)
                                 Hash Cond: (payments.id = payments_1.id)
                                 ->  Parallel Seq Scan on payments  (cost=0.00..35960.56 rows=784456 width=61) (
actual time=0.056..3110.479 rows=627565 loops=3)
                                 ->  Parallel Hash  (cost=148259.10..148259.10 rows=394101 width=8) (actual time
=16233.277..16233.590 rows=309835 loops=3)
                                       Buckets: 131072  Batches: 16  Memory Usage: 3360kB
                                       ->  Parallel Hash Join  (cost=96764.85..148259.10 rows=394101 width=8) (a
ctual time=13136.651..14693.417 rows=309835 loops=3)
                                             Hash Cond: (payments_1.donation_id = donations.id)
                                             ->  Parallel Hash Join  (cost=67156.42..112432.59 rows=406880 width
=12) (actual time=7591.899..9196.321 rows=311337 loops=3)
                                                   Hash Cond: (payments_1.supporter_id = supporters_1.id)
                                                   ->  Parallel Seq Scan on payments payments_1  (cost=0.00..379
21.70 rows=429516 width=16) (actual time=0.076..1739.948 rows=343035 loops=3)
                                                         Filter: (nonprofit_id = 4341)
                                                         Rows Removed by Filter: 284530
                                                   ->  Parallel Hash  (cost=58627.52..58627.52 rows=519832 width
=4) (actual time=4125.461..4125.480 rows=416677 loops=3)
                                                         Buckets: 131072  Batches: 32  Memory Usage: 2592kB
                                                         ->  Parallel Seq Scan on supporters supporters_1  (cost
=0.00..58627.52 rows=519832 width=4) (actual time=0.065..2066.998 rows=416677 loops=3)
                                                               Filter: (NOT anonymous)
                                                               Rows Removed by Filter: 22324
                                             ->  Parallel Hash  (cost=24671.07..24671.07 rows=300909 width=4) (a
ctual time=2379.479..2379.493 rows=240799 loops=3)
                                                   Buckets: 131072  Batches: 16  Memory Usage: 2848kB
                                                   ->  Parallel Seq Scan on donations  (cost=0.00..24671.07 rows
=300909 width=4) (actual time=0.063..1191.882 rows=240799 loops=3)
                                                         Filter: (NOT anonymous)
                                                         Rows Removed by Filter: 1927
                           ->  Parallel Hash  (cost=58627.52..58627.52 rows=548752 width=40) (actual time=4382.1
26..4382.141 rows=439001 loops=3)
                                 Buckets: 65536  Batches: 32  Memory Usage: 3296kB
                                 ->  Parallel Seq Scan on supporters  (cost=0.00..58627.52 rows=548752 width=40) (actual time=0.040..2190.589 rows=439001 loops=3)
                     ->  Hash  (cost=164.94..164.94 rows=1594 width=21) (actual time=46.148..46.162 rows=1594 loops=3)
                           Buckets: 2048  Batches: 1  Memory Usage: 89kB
                           ->  Seq Scan on nonprofits  (cost=0.00..164.94 rows=1594 width=21) (actual time=29.881..38.148 rows=1594 loops=3)
   ->  Materialize  (cost=0.28..4.30 rows=1 width=4) (actual time=0.005..0.010 rows=1 loops=929505)
         ->  Index Only Scan using npos_pkey on nonprofits nonprofits_1  (cost=0.28..4.29 rows=1 width=4) (actual time=0.018..0.028 rows=1 loops=1)
               Index Cond: (id = 4341)
               Heap Fetches: 0
 Planning Time: 1.730 ms
 JIT:
   Functions: 140
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 12.673 ms, Inlining 0.000 ms, Optimization 4.498 ms, Emission 85.948 ms, Total 103.119 ms
 Execution Time: 76205.459 ms
```

And this is the query:

```sql
EXPLAIN ANALYZE SELECT
  payments.kind,
  payments.towards,
  payments.id AS id,
  supporters.name,
  supporters.email,
  payments.gross_amount,
  timezone(
    COALESCE(nonprofits.timezone, 'UTC'),
    timezone('UTC', payments.date)
  ) as date
FROM
  payments
  INNER JOIN supporters ON supporters.id = payments.supporter_id
  INNER JOIN nonprofits ON nonprofits.id = payments.nonprofit_id
  LEFT OUTER JOIN donations ON donations.id = payments.donation_id
  JOIN (
    SELECT
      payments.id
    FROM
      payments
      INNER JOIN supporters ON supporters.id = payments.supporter_id
      INNER JOIN nonprofits ON nonprofits.id = payments.nonprofit_id
      LEFT OUTER JOIN (
        SELECT
          donations.*
        FROM
          donations
      ) AS donations ON donations.id = payments.donation_id
    WHERE
      (
        payments.nonprofit_id = <nonprofit_id>
        AND donations.anonymous = 'false'
        AND supporters.anonymous = 'false'
      )
  ) AS "filtered_payments" ON payments.id = filtered_payments.id
ORDER BY
  payments.date DESC;
```

clarissalimab commented:
Based on your feedback on the UI, I changed the text and added the label, but I think it looked better being placed before the kind of payments. What do you think?

![image](https://user-images.githubusercontent.com/15739610/135862792-5d6d353b-f271-480f-a563-3bada282c93f.png)

clarissalimab commented on lib/query/query_payments.rb:
```diff
@@ -144,8 +144,10 @@ def self.select_to_filter_search(npo_id, query)
     end
+
+    inner_supporters_search = Qexpr.new.select('supporters.*').from('supporters').where('NOT supporters.deleted')
     expr = Qexpr.new.select('payments.id').from('payments')
-          .inner_join('supporters', "supporters.id=payments.supporter_id")
+          .inner_join(inner_supporters_search.as('supporters'), "supporters.id=payments.supporter_id")
```
Does it make sense with what you said about not including deleted supporters on the result?

wwahammy replied:
Was it previously including deleted supporters in the result?

wwahammy replied:
Oops, I remember this now: you need to have payments from deleted supporters as well so this change should not be made.

After all, consider a charge that was made and then supporter was deleted. The nonprofit would want that record to show as a payment (yes, we should have a better way to handling "deleted" supporters).

Please put a comment there explaining why we don't want only deleted supporters there, otherwise we may make this mistake again :smile:

clarissalimab commented on db/migrate/20211001194346_add_index_to_anonymous_on_supporter.rb:
```diff
@@ -0,0 +1,5 @@
+class AddIndexToAnonymousOnSupporter < ActiveRecord::Migration
+  def change
+    add_index :supporters, [:anonymous, :nonprofit_id]
```
Do we want to keep it this way?

wwahammy replied:
I think so?

clarissalimab commented:
I'm having problems with the tests where the Dontation is being created with the anonymous field being nil, even with the default defined as false already. I tried to reset the database to make sure everything was updated, tried to run `rake db:migrate RAILS_ENV=test` but nothing worked :confused:

wwahammy commented on lib/insert/insert_donation.rb:
```diff
@@ -173,7 +173,7 @@ def self.insert_payment(kind, fee_total, donation_id, data)
     d = Donation.new
     d.date = data['date']
-    d.anonymous = data['anonymous']
+    d.anonymous = data['anonymous'].nil? ? false : true
```
We should move this into a `:before_save` on Donation too.

clarissalimab replied:
It's already there, I was debugging and committed this by accident, sorry!

wwahammy reviewed (changes requested):
Looking pretty good. One thing we should test is whether payment updating works properly, i.e. they use the `Donation` class when we've moved nudging nil anonymous fields to false to inside the class.

clarissalimab commented:
I couldn't find anywhere that a payment is being updated or saved that also changes `donation.anonymous`, I did a smoke test and it worked properly.

wwahammy commented:
I'm going to look into testing this now.

wwahammy reviewed (approved):
I think this looks good. We can discuss merging and deploying on a private ticket.
